### PR TITLE
Use different relation for `instanceof` type guards

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Implements #4107

### CODE REVIEW

File: tests/baselines/reference/typeGuardsWithInstanceOfByConstructorSignature.errors.txt [hunk 1]
```diff
@@ -84,11 +82,9 @@ tests/cases/conformance/expressions/typeGuards/typeGuardsWithInstanceOfByConstru
         obj5.foo;
         obj5.c;
         obj5.bar1;
-             ~~~~
-!!! error TS2339: Property 'bar1' does not exist on type 'C1 | C2'.
```
Comment (RyanCavanaugh) [inline tests/baselines/reference/typeGuardsWithInstanceOfByConstructorSignature.errors.txt]: I believe this is an improvement -- `obj5` was of type `C1 | A`, so a narrowing to `instanceof C` should not add `C2` back into the mix.

File: tests/cases/compiler/instanceOfAssignability.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
+interface Base {
+	foo: string|number;
```
Comment (DanielRosenwasser) [inline tests/cases/compiler/instanceOfAssignability.ts]: Can you please change your editor's tab settings? It makes for a more consistent diff.

File: tests/cases/compiler/instanceOfAssignability.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+		let y = x;
+	}
+}
+
+function fn5(x: Derived1) {
+	if(x instanceof Derived2) {
+		// 1.5: y: Derived1
+		// Want: ???
+		let y = x;
+	}
```
Comment (DanielRosenwasser) [inline tests/cases/compiler/instanceOfAssignability.ts]: Can you please change your editor's tab settings? It makes for a more consistent diff.

Comment (RyanCavanaugh) [inline tests/cases/compiler/instanceOfAssignability.ts]: Sometimes I use notepad `¯\_(ツ)_/¯`. Homogeneity in our test suite is a bad thing, not a good thing. How are we going to notice if one of our plugins starts totally barfing on `\t`s?

You can use `?ts=4` in the GitHub differ if the tab size is too distracting.

File: tests/baselines/reference/typeGuardsWithInstanceOfByConstructorSignature.errors.txt (hunk 1 above)
Comment (JsonFreeman) [inline tests/baselines/reference/typeGuardsWithInstanceOfByConstructorSignature.errors.txt]: Agreed.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -6207,14 +6207,20 @@ namespace ts {
...
-                // Narrow to the target type if it's a subtype of the current type
-                if (isTypeSubtypeOf(narrowedTypeCandidate, originalType)) {
-                    return narrowedTypeCandidate;
-                }
-                // If the current type is a union type, remove all constituents that aren't subtypes of the target.
+                // If the current type is a union type, remove all constituents that aren't assignable to target. If that produces
+                // 0 candidates, fall back to the assignability check
                 if (originalType.flags & TypeFlags.Union) {
-                    return getUnionType(filter((<UnionType>originalType).types, t => isTypeSubtypeOf(t, narrowedTypeCandidate)));
+                    let assignableConsituents = filter((<UnionType>originalType).types, t => isTypeAssignableTo(t, narrowedTypeCandidate));
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Fix spelling `assignableConstituents`

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -6207,14 +6207,20 @@ namespace ts {
...
+                    if (assignableConsituents.length) {
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: I think this length check makes sense. It is never helpful to give the user empty object type just because we ran out of constituents.

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -6207,14 +6207,20 @@ namespace ts {
...
+                        return getUnionType(assignableConsituents);
+                    }
                 }
+                
+                if (isTypeAssignableTo(narrowedTypeCandidate, originalType)) {
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Curious, why did you move this assignability check to the bottom?

**Comment by ahejlsberg** [comment]
:+1:
